services/job_runner.go

```diff
@@ -22,38 +23,48 @@ type JobRunner interface {
 }
 
 type jobRunner struct {
-	closed       bool
-	closer       chan struct{}
-	closingMutex sync.Mutex
-	store        *store.Store
-	workerMutex  sync.Mutex
-	workers      map[string]chan store.RunRequest
+	started     bool
+	done        chan struct{}
+	bootMutex   sync.Mutex
+	store       *store.Store
+	workerMutex sync.RWMutex
+	workers     map[string]chan store.RunRequest
+	wg          sync.WaitGroup
 }
 
 // NewJobRunner initializes a JobRunner.
 func NewJobRunner(str *store.Store) JobRunner {
 	return &jobRunner{
-		closer:  make(chan struct{}),
 		store:   str,
 		workers: make(map[string]chan store.RunRequest),
 	}
 }
 
 // Start reinitializes runs and starts the execution of the store's runs.
 func (rm *jobRunner) Start() error {
+	rm.bootMutex.Lock()
+	defer rm.bootMutex.Unlock()
+
+	if rm.started {
+		return errors.New("JobRunner already started")
+	}
+	rm.done = make(chan struct{})
+	rm.started = true
 	go rm.demultiplexRuns()
 	return rm.resumeSleepingRuns()
 }
 
 // Stop closes all open worker channels.
 func (rm *jobRunner) Stop() {
-	rm.closingMutex.Lock()
-	defer rm.closingMutex.Unlock()
+	rm.bootMutex.Lock()
+	defer rm.bootMutex.Unlock()
 
-	if !rm.closed {
-		rm.closed = true
-		close(rm.closer)
+	if !rm.started {
```

Comment:
Seems like `bootMutex` and `started` could be an atomic rather than a mutex + bool?

Reply:
The bootMutex is protecting more than just the bool, conceptually I see it as protecting the whole `start()` method. The creation and destruction of `rm.done` is another example of something we don't want accessed in parallel. But perhaps I'm being overprotective.